keras/applications/inception_resnet_v2.py

```diff
@@ -0,0 +1,383 @@
+            will have `block_type='block35', block_idx=0`, ane the layer names will have
+            a common prefix `'block35_0'`.
+        activation: activation function to use at the end of the block
+            (see [activations](../activations.md)).
+            When `activation=None`, no activation is applied
+            (i.e., "linear" activation: `a(x) = x`).
+
+    # Returns
+        Output tensor for the block.
+
+    # Raises
+        ValueError: if `block_type` is not one of `'block35'`,
+            `'block17'` or `'block8'`.
+    """
+    if block_type == 'block35':
+        branch_0 = conv2d_bn(x, 32, 1)
+        branch_1 = conv2d_bn(x, 32, 1)
+        branch_1 = conv2d_bn(branch_1, 32, 3)
+        branch_2 = conv2d_bn(x, 32, 1)
+        branch_2 = conv2d_bn(branch_2, 48, 3)
+        branch_2 = conv2d_bn(branch_2, 64, 3)
+        branches = [branch_0, branch_1, branch_2]
+    elif block_type == 'block17':
+        branch_0 = conv2d_bn(x, 192, 1)
+        branch_1 = conv2d_bn(x, 128, 1)
+        branch_1 = conv2d_bn(branch_1, 160, [1, 7])
+        branch_1 = conv2d_bn(branch_1, 192, [7, 1])
+        branches = [branch_0, branch_1]
+    elif block_type == 'block8':
+        branch_0 = conv2d_bn(x, 192, 1)
+        branch_1 = conv2d_bn(x, 192, 1)
+        branch_1 = conv2d_bn(branch_1, 224, [1, 3])
+        branch_1 = conv2d_bn(branch_1, 256, [3, 1])
+        branches = [branch_0, branch_1]
+    else:
+        raise ValueError('Unknown Inception-ResNet block type. '
+                         'Expects "block35", "block17" or "block8", '
+                         'but got: ' + str(block_type))
+
+    block_name = block_type + '_' + str(block_idx)
+    channel_axis = 1 if K.image_data_format() == 'channels_first' else 3
+    mixed = Concatenate(axis=channel_axis, name=block_name + '_mixed')(branches)
+    up = conv2d_bn(mixed,
+                   K.int_shape(x)[channel_axis],
+                   1,
+                   activation=None,
+                   use_bias=True,
+                   name=block_name + '_conv')
+
+    x = Lambda(lambda inputs, scale: inputs[0] + inputs[1] * scale,
+               output_shape=K.int_shape(x)[1:],
+               arguments={'scale': scale},
+               name=block_name)([x, up])
+    if activation is not None:
+        x = Activation(activation, name=block_name + '_ac')(x)
+    return x
+
+
+def InceptionResNetV2(include_top=True,
+                      weights='imagenet',
+                      input_tensor=None,
+                      input_shape=None,
+                      pooling=None,
+                      classes=1000):
+    """Instantiates the Inception-ResNet v2 architecture.
+
+    Optionally loads weights pre-trained on ImageNet.
+    Note that when using TensorFlow, for best performance you should
+    set `"image_data_format": "channels_last"` in your Keras config
+    at `~/.keras/keras.json`.
+
+    The model and the weights are compatible with both TensorFlow and Theano
```

Comment:
Sure about theano? I only see tf at line 370.

Reply:
I've tested the model on sample images and the result seems reasonable. I think weight conversion from TF to TH is handled in `model.load_weights()`, since backend information is included in the weight files.

Reply:
Correct, in theory it should work with both. What results do you see?

Reply:
For example, given a random truck image, the prediction is:

With TensorFlow:

    [[('n03417042', 'garbage_truck', 0.77079743),
      ('n03126707', 'crane', 0.05554872),
      ('n03796401', 'moving_van', 0.051750395),
      ('n04252225', 'snowplow', 0.028239522),
      ('n04467665', 'trailer_truck', 0.01623963)]]

With Theano:

    [[('n03417042', 'garbage_truck', 0.76859063),
      ('n03126707', 'crane', 0.059462097),
      ('n03796401', 'moving_van', 0.054951463),
      ('n04252225', 'snowplow', 0.027199779),
      ('n04467665', 'trailer_truck', 0.015053505)]]

The probabilities are not exactly the same, but the order is correct (also tested on `InceptionV3` and `ResNet50`, and observed similar results). Changing the `image_data_format` does not affect the probabilities at all.

Reply:
The discrepancy is likely due to a known issue with how Theano and TF use different implementations for unpadded pooling operations. It's fine.